certora/specs/YieldStreams.spec

```diff
@@ -0,0 +1,800 @@
+        _maxLossOnOpenTolerance);
+
+    require ownerOf(streamIds[0]) == dave;
+
+    // Postconditions
+    assert streamIds.length == 1 &&
+     streamIds[0] == initialNextStreamId &&
+     to_mathint(nextStreamId()) == initialNextStreamId + 1;
+    assert to_mathint(receiverTotalPrincipal(bob)) == initialBobPrincipal + _principal &&
+     receiverPrincipal(bob, streamIds[0]) == _principal && vault.balanceOf(currentContract) == shares;
+}
+
+/**
+ * @Rule Integrity property for the `depositAndOpenMultipleUsingPermit` function
+ * @Category High
+ * @Description Ensures that the `depositAndOpenMultipleUsingPermit` function opens multiple yield streams using ERC20 permit for approval, allocating the underlying asset as principal.
+ */
+rule integrity_of_depositAndOpenMultipleUsingPermit(address dave, address bob, uint256 principal,
+        address[] receivers,
+        uint256[] allocations,
+        uint256 deadline,
+        uint8 v,
+        bytes32 r,
+        bytes32 s) {
+    env e;
+    require dave != bob && dave == e.msg.sender;
+    require principal > 0;
+    require to_mathint(deadline) == e.block.timestamp + 1;
+    require asset.balanceOf(dave) == principal;
+
+    require(asset.balanceOf(dave) >= principal && asset.allowance(dave, currentContract) >= principal);
+
+    require to_mathint(asset.balanceOf(vault)) == 2 * principal;
+ 
+    require receivers.length == allocations.length && allocations.length == 1;
+    require receivers[0] == bob && allocations[0] == principal;
+
+    uint256 shares = vault.deposit(e, principal, currentContract);
+    asset.approve(e, currentContract, shares);
+
+    // Capture the initial state
+    uint256 initialNextStreamId = nextStreamId();
+    uint256 initialBobShares = receiverTotalShares(bob);
+    uint256 initialBobPrincipal = receiverTotalPrincipal(bob);
+
+    // Call the depositAndOpenMultipleUsingPermit function
+    uint256[] streamIds = depositAndOpenMultipleUsingPermit(
+        e,
+        principal,
+        receivers,
+        allocations,
+        0,
+        deadline,
+        v,
+        r,
+        s
+    );
+
+    require ownerOf(streamIds[0]) == dave;
+
+    // Postconditions
+    assert streamIds.length == 1 &&
+     streamIds[0] == initialNextStreamId &&
+     to_mathint(nextStreamId()) == initialNextStreamId + 1;
+    assert to_mathint(receiverTotalPrincipal(bob)) == initialBobPrincipal + principal &&
+     receiverPrincipal(bob, streamIds[0]) == principal && vault.balanceOf(currentContract) == shares;
+}
+
+
+/**
+ * @Rule Integrity property for the `topUp` function
+ * @Category High
+ * @Description Ensures that the `topUp` function adds the specified shares to the existing yield stream and update the contract state accordingly, if the caller is the owner of the stream.
+ */
+
+rule integrity_of_topUp(uint256 _streamId, uint256 _shares) {
+    env e;
+    address streamer = e.msg.sender;
+    address receiver = streamIdToReceiver(_streamId);
+    uint256 principal;
```

Comment:
It is easier to read if moving the declaration to the assignment, i.e.,
`uint256 principal = topUp(e, _streamId, _shares);`

Reply:
Updated